core/new-gui/src/app/workspace/component/visualization-panel-content/visualization-panel-content.component.ts

```diff
@@ -102,6 +119,95 @@ export class VisualizationPanelContentComponent implements AfterViewInit, OnDest
       case ChartType.SPLINE:
         this.generateChart();
         break;
+      case ChartType.SPATIAL_SCATTERPLOT:
+        this.generateSpatialScatterplot();
+        break;
+      case ChartType.SCATTERPLOT:
+        this.generateSimpleScatterplot();
+        break;
+    }
+  }
+
+  generateSimpleScatterplot() {
+    if (this.c3ChartElement) {
+      this.c3ChartElement.destroy();
+    }
+    const result = this.data as Array<Record<string, Primitive>>;
+    const xLabel: string = Object.keys(result[0])[0];
+    const yLabel: string = Object.keys(result[0])[1];
+
+    this.c3ChartElement = c3.generate({
+      size: {
+        height: VisualizationPanelContentComponent.HEIGHT,
+        width: VisualizationPanelContentComponent.WIDTH
+      },
+      data: {
+        json: result,
+        keys: {
+          x: xLabel,
+          value: [yLabel]
+        },
+        type: this.chartType as c3.ChartType
+      },
+      axis: {
+        x: {
+          label: xLabel,
+          tick: {
+            fit: true
+          }
+        },
+        y: {
+          label: yLabel
+        }
+      },
+      bindto: VisualizationPanelContentComponent.CHART_ID
+    });
+  }
+
+  generateSpatialScatterplot() {
+    if (this.map === undefined) {
+      this.initMap();
+    }
+    this.map?.on('styledata', () => {
+      this.addScatterLayer();
+    });
+  }
+
+  initMap() {
+    /* mapbox object with default configuration */
+    this.map = new mapboxgl.Map({
+      container: VisualizationPanelContentComponent.MAP_CONTAINER,
+      style: 'mapbox://styles/mapbox/light-v9',
+      center: [-96.35, 39.5],
+      zoom: 3,
+      maxZoom: 17,
+      minZoom: 0
+    });
+  }
+
+  addScatterLayer() {
+    if (this.map !== undefined) {
+      if (this.map?.getLayer('scatter')) {
+        this.map?.removeLayer('scatter');
+      }
+      const props: ScatterplotLayerProps<any> = {
+        opacity: 0.8,
+        filled: true,
+        radiusScale: 100,
+        radiusMinPixels: 1,
+        radiusMaxPixels: 25,
+        getPosition: (d: { longitude: any; latitude: any; }) => [d.longitude, d.latitude],
+        getFillColor: [57, 73, 171]
+      };
```

Comment:
this block seems to be a static constant property. can we define it once in the outer scope instead?

Reply:
no, it depends on the data `d` can't be static

Reply:
I don't get it. Isn't the getPosition a callback function?